backends/rapidpro/msg.go

```diff
@@ -262,63 +265,61 @@ func (b *backend) flushMsgFile(filename string, contents []byte) error {
 // Deduping utility methods
 //-----------------------------------------------------------------------------
 
-// checkMsgSeen tries to look up whether a msg with the fingerprint passed in was seen in window or prevWindow. If
-// found returns the UUID of that msg, if not returns empty string
+// checks to see if this message has already been seen and if so returns its UUID
 func (b *backend) checkMsgSeen(msg *DBMsg) courier.MsgUUID {
 	rc := b.redisPool.Get()
 	defer rc.Close()
 
-	uuidAndText, _ := b.seenMsgs.Get(rc, msg.fingerprint(false))
+	// if we have an external id use that
+	if msg.ExternalID_ != "" {
+		fingerprint := fmt.Sprintf("%s|%s|%s", msg.Channel().UUID(), msg.URN().Identity(), msg.ExternalID())
 
-	// if so, test whether the text it the same
-	if uuidAndText != "" {
-		prevText := uuidAndText[37:]
+		uuid, _ := b.seenExternalIDs.Get(rc, fingerprint)
 
-		// if it is the same, return the UUID
-		if prevText == msg.Text() {
-			return courier.MsgUUID(uuidAndText[:36])
+		if uuid != "" {
+			return courier.MsgUUID(uuid)
 		}
-	}
-	return courier.NilMsgUUID
-}
+	} else {
+		// otherwise de-dup based on text received from that channel+urn since last send
+		fingerprint := fmt.Sprintf("%s|%s", msg.Channel().UUID(), msg.URN().Identity())
 
-// writeMsgSeen writes that the message with the passed in fingerprint and UUID was seen in the
-// passed in window
-func (b *backend) writeMsgSeen(msg *DBMsg) {
-	rc := b.redisPool.Get()
-	defer rc.Close()
+		uuidAndText, _ := b.seenMsgs.Get(rc, fingerprint)
 
-	b.seenMsgs.Set(rc, msg.fingerprint(false), fmt.Sprintf("%s|%s", msg.UUID(), msg.Text()))
-}
+		// if we have seen a message from this channel+urn check text too
+		if uuidAndText != "" {
+			prevText := uuidAndText[37:]
 
-// clearMsgSeen clears our seen incoming messages for the passed in channel and URN
-func (b *backend) clearMsgSeen(rc redis.Conn, msg *DBMsg) {
-	b.seenMsgs.Remove(rc, msg.fingerprint(false))
+			// if it is the same, return the UUID
+			if prevText == msg.Text() {
+				return courier.MsgUUID(uuidAndText[:36])
+			}
+		}
+	}
```

Comment:
I understand the issue was about Telegram when we started this that gives us external IDs, but about making sure that any channel type that will send 2 messages without text but different attachments are also save as too messages, I think here, they will not as long as the are sent with the same timestamp

Reply:
Yeah tho I can't imagine there are many channels that send attachments but don't give us external ids.

We can merge this and then you can add your previous changes to use a hash of text+attachments.

Reply:
Right, that is not something that is likely to happen it is just that I had found that is could be an issue in theory 

I can add support for that after the merge